clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,334 @@
+        this.ledger = new StateLedger();
+    }
+
+    public void init(List<KafkaMetric> metrics) {
+        ledger.init(metrics);
+    }
+
+    /**
+     * This is called whenever a metric is updated or added
+     */
+    public void metricChange(KafkaMetric metric) {
+        ledger.metricChange(metric);
+    }
+
+    /**
+     * This is called whenever a metric is removed
+     */
+    public void metricRemoval(KafkaMetric metric) {
+        ledger.metricRemoval(metric);
+    }
+
+    // Visible for testing
+    Set<MetricKey> getTrackedMetrics() {
+        return ledger.metricMap.keySet();
+    }
+
+    @Override
+    public void collect(MetricsEmitter metricsEmitter) {
+        for (Map.Entry<MetricKey, KafkaMetric> entry : ledger.getMetrics()) {
+            MetricKey metricKey = entry.getKey();
+            KafkaMetric metric = entry.getValue();
+
+            try {
+                collectMetric(metricsEmitter, metricKey, metric);
+            } catch (Exception e) {
+                // catch and log to continue processing remaining metrics
+                log.error("Error processing Kafka metric {}", metricKey, e);
+            }
+        }
+    }
+
+    protected void collectMetric(MetricsEmitter metricsEmitter, MetricKey metricKey, KafkaMetric metric) {
+        Object metricValue;
+
+        try {
+            metricValue = metric.metricValue();
+        } catch (Exception e) {
+            // If an exception occurs when retrieving value, log warning and continue to process the rest of metrics
+            log.warn("Failed to retrieve metric value {}", metricKey.name(), e);
+            return;
+        }
+
+        Instant now = Instant.ofEpochMilli(time.milliseconds());
+        if (isMeasurable(metric)) {
+            Measurable measurable = metric.measurable();
+            Double value = (Double) metricValue;
+
+            if (measurable instanceof WindowedCount || measurable instanceof CumulativeSum) {
+                collectSum(metricKey, value, metricsEmitter, now);
+            } else {
+                collectGauge(metricKey, value, metricsEmitter, now);
+            }
+        } else {
+            // It is non-measurable Gauge metric.
+            // Collect the metric only if its value is a number.
+            if (metricValue instanceof Number) {
+                Number value = (Number) metricValue;
+                collectGauge(metricKey, value, metricsEmitter, now);
+            } else {
+                // skip non-measurable metrics
+                log.debug("Skipping non-measurable gauge metric {}", metricKey.name());
+            }
+        }
+    }
+
+    private void collectSum(MetricKey metricKey, double value, MetricsEmitter metricsEmitter, Instant timestamp) {
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        if (metricsEmitter.shouldEmitDeltaMetrics()) {
+            InstantAndValue<Double> instantAndValue = ledger.delta(metricKey, timestamp, value);
```

Comment:
if the aggregation temporality can change between collection cycles then this would not give correct if temporality switches back from cumulative to delta, since the ledger wouldn't have been updated for a long time.

While this is a pathological case, I think it would be simpler if aggregation temporality was fixed on startup so that we don't have to worry about correctness on every collection.

@AndrewJSchofield do we know what guarantees the broker provides with respect to temporality? It would seem unlikely it would need to change unless there is re-connection to the broker.

Reply:
For AK 3.7, the broker will only use delta temporality and there's no interface for specifying anything else. So, the the least, we would need a follow-on KIP to define how the administrator can control the temporality and describe what happens when it changes.

Reply:
in that case I would suggest we don't try to support both yet, and re-introduce that code when we are ready. I think that will simplify things now and avoid baking in incorrect assumption into the codebase at this stage.

Reply:
Having talked this over with @apoorvmittal10, we think it is viable to reset the ledger in the event that the temporality changes. That is a nice clean behaviour in the client code which is easily testable and understandable.

We have various enhancements in mind for improving how client metrics are administered on the broker end, but they can be accomplished without client changes, which is important because upgrading clients is such a lengthy process.

In summary, I think we should support both temporalities in the client and the transition should be clean and therefore easily testable.

Reply:
@xvrl @AndrewJSchofield I have added this jira for a follow up PR: https://issues.apache.org/jira/browse/KAFKA-15877. I ll raise the reset of ledger when temporality changes from Cumulative to Delta.

Reply:
I have addressed the concern in same PR and have added way to reset the value tracker. I have added tests too to verify same. Resolving the comment.